ctest-next/src/template.rs

```diff
@@ -2,78 +2,185 @@ use askama::Template;
-    translator: Translator,
-    ffi_items: &'a FfiItems,
-    generator: &'a TestGenerator,
+pub(crate) struct CTestTemplate {
+    pub(crate) template: TestTemplate,
+    pub(crate) headers: Vec<String>,
 }
 
-impl<'a> RustTestTemplate<'a> {
-    /// Create a new test template to test the given items.
-    pub(crate) fn new(ffi_items: &'a FfiItems, generator: &'a TestGenerator) -> Self {
-        Self {
-            ffi_items,
-            generator,
-        }
+impl CTestTemplate {
+    pub(crate) fn new(
+        ffi_items: &FfiItems,
+        generator: &TestGenerator,
+    ) -> Result<Self, TranslationError> {
+        Ok(Self {
+            template: TestTemplate::new(ffi_items, generator)?,
+            headers: generator.headers.clone(),
+        })
     }
 }
 
-impl<'a> CTestTemplate<'a> {
-    /// Create a new test template to test the given items.
-    pub(crate) fn new(ffi_items: &'a FfiItems, generator: &'a TestGenerator) -> Self {
-        Self {
+/// Stores all information necessary for generation of tests for all items.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct TestTemplate {
+    pub(crate) const_cstr_tests: Vec<TestCstr>,
+    pub(crate) const_tests: Vec<TestConst>,
+    pub(crate) test_idents: Vec<BoxStr>,
+}
+
+impl TestTemplate {
+    pub(crate) fn new(
+        ffi_items: &FfiItems,
+        generator: &TestGenerator,
+    ) -> Result<Self, TranslationError> {
+        let helper = TranslateHelper {
             ffi_items,
-            translator: Translator::new(),
             generator,
+            translator: Translator::new(),
+        };
+
+        /* Figure out which tests are to be generated. */
+        // FIXME(ctest): Populate more test information, maybe extract into separate methods.
+        // The workflow would be to create a struct that stores information for the new test,
+        // and populating that struct here, so that the also things that have to be added to
+        // the test templates are the new tests parameterized by that struct.
+
+        let mut const_tests = vec![];
+        let mut const_cstr_tests = vec![];
+        for constant in ffi_items.constants() {
+            if let syn::Type::Ptr(ptr) = &constant.ty {
+                let is_const_c_char_ptr = matches!(
+                    &*ptr.elem,
+                    syn::Type::Path(path)
+                        if path.path.segments.last().unwrap().ident == "c_char"
+                        && ptr.mutability.is_none()
+                );
+                if is_const_c_char_ptr {
+                    let item = TestCstr {
+                        test_ident: cstr_test_ident(constant.ident()),
+                        rust_ident: constant.ident().into(),
+                        c_ident: helper.c_ident(constant).into(),
+                        c_type: helper.c_type(constant)?.into(),
+                    };
+                    const_cstr_tests.push(item)
+                }
+            } else {
```

Comment:
This doesn't add any test in the case where it is a pointer but not a `CStr`. Either make the top-level `if` also check for the pointer type so the nested `if` isn't needed:

```rust
matches!(
    &constant.ty,
    syn::Type::Ptr(syn::Type::Path(path))
        if path.path.segments.last().unwrap().ident == "c_char"
        && ptr.mutability.is_none()
);
```

Or bump the MSRV by 1 again and use let chains to do the same thing but slightly cleaner.